[LoopVectorize] Enable shuffle padding for masked interleaved accesses

huhu233 opened the pull request:
```
typedef struct {
    float x;
    float y;
    float z;
} patic;

for (int i = 0; i < num; i++) {
  ps[i].x = factor * ps[i].x;
  ps[i].y = factor * ps[i].y;
}
```

This patch pads the gap of the interleave store group to eliminate masked.store, which helps to generate better code in Interleaved Access pass, as shown,
```
%wide.vec = load <12 x float>; 0,1,2,3,...,11
%shuffle1 = shuffle %wide.vec, poison, <0, 3, 6, 9> ; 0,3,6,9 
%shuffle2 = shuffle %wide.vec, poison, <1, 4, 7, 10> ; 1,4,7,10 
%padded = shuffle %wide.vec, poison, <2, 5, 8, 11> ; 2,5,8,11 

%concate1 = shuffle %op1, %op2, <0, 1, ..., 7> ; 0,3,6,9,1,4,7,10 
%concate2 = shuffle %padded, poison, <0, 1, ..., 3, undef, undef, undef, undef> ; 2,5,8,11,poison,...,poison
%concateFinal = shuffle %concate1, %concate2, <0, 4, 8, 1, 5, 9, 2, 6, 10, 3, 7, 11> ; 0,1,2,3,...,11
 store <12 x float> %concateFinal
```
This patch adds some restrictions for shuffle padding, that is target interleaved store groups should have matched interleaved load groups, which means,
1. The value operand of StoreInst should comes from LoadInst.
2. The store group and the load group accesses the same underlying object.

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/sve-structured-store.ll:
```diff
@@ -0,0 +1,279 @@
+; RUN: opt -mtriple=aarch64 -mattr=+sve -S -passes=loop-vectorize,interleaved-access  -enable-masked-interleaved-mem-accesses=true -sve-gather-overhead=10 -sve-scatter-overhead=10  -enable-shuffle-padding=true -force-vector-width=16 -aarch64-sve-vector-bits-min=512 < %s -o - | FileCheck %s --check-prefixes=ENABLE-SHUFFLE-PADDING
```
The `-sve-gather-overhead=10 -sve-scatter-overhead=10` flags aren't needed because they are the default already I think. Also, the `-aarch64-sve-vector-bits-min=512` flag contradicts the `vscale_range(2,2)` attribute (i.e. min=256 bits).

huhu233 replied:
Fixed, remove unnecessary tokens.

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/sve-structured-store.ll:
```diff
@@ -0,0 +1,279 @@
+;
+; Function Attrs: argmemonly mustprogress nofree norecurse nosync nounwind uwtable vscale_range(2,2)
+define void @test(i32 noundef %num, ptr nocapture noundef %ps) {
+; ENABLE-SHUFFLE-PADDING-LABEL: @test(
```
The ENABLE and DISABLE check lines are identical so I don't know what you're trying to show here that's different? What effect does your patch have on these tests? At the moment it's hard to know the impact of your code change.

huhu233 replied:
Hi, @david-arm, check lines for ENABLE and DISABLE are different. When the shuffle padding is enabled, the interleaved store will be transformed into `llvm.aarch64.sve.st3`.  And if the feature is disabled, there is no `sve.st3`, and I use `ISABLE-SHUFFLE-PADDING-NOT` here.

david-arm replied:
My apologies, I see that now! In that case, I think it's worth having more testing than just the existence of st3 in the IR for the ENABLE-SHUFFLE case. I believe it's important to show the input to the store as well, i.e. how the stored value is generated.

huhu233 replied:
Done, but it seems that the file has become very large...

david-arm commented on llvm/test/Transforms/LoopVectorize/AArch64/sve-structured-store.ll:
```diff
@@ -0,0 +1,279 @@
+;   ps[i].y = factor * ps[i].y;
+; }
+;
+; Function Attrs: argmemonly mustprogress nofree norecurse nosync nounwind uwtable vscale_range(2,2)
```
Do you even need to specify a min vscale > 1 here? The most common case users will hit with SVE are `vscale_range(1,16)`.

huhu233 replied:
Fixed, thanks!

Commit:
[LoopVectorize] Enable shuffle padding for masked interleaved accesses

typedef struct {
    float x;
    float y;
    float z;
} patic;

for (int i = 0; i < num; i++) {
  ps[i].x = factor * ps[i].x;
  ps[i].y = factor * ps[i].y;
}

This patch pads the gap of the interleave store group to eliminate
masked.store, which helps to generate better code in Interleaved Access pass,
as shown,

%wide.vec = load <12 x float>; 0,1,2,3,...,11
%shuffle1 = shuffle %wide.vec, poison, <0, 3, 6, 9> ; 0,3,6,9
%shuffle2 = shuffle %wide.vec, poison, <1, 4, 7, 10> ; 1,4,7,10
%padded = shuffle %wide.vec, poison, <2, 5, 8, 11> ; 2,5,8,11

%concate1 = shuffle %op1, %op2, <0, 1, ..., 7> ; 0,3,6,9,1,4,7,10
%concate2 = shuffle %padded, poison,
            <0, 1, ..., 3, undef, undef, undef, undef> ; 2,5,8,11,poison,...,poison
%concateFinal = shuffle %concate1, %concate2,
                <0, 4, 8, 1, 5, 9, 2, 6, 10, 3, 7, 11> ; 0,1,2,3,...,11
store <12 x float> %concateFinal

This patch adds some restrictions for shuffle padding, that is
target interleave store groups should have matched interleave
load groups, which means,
1. The value operand of StoreInst should comes from LoadInst.
2. The store group and the load group accesses the same strcut
memory.

nikolaypanchenko commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -2672,6 +2707,24 @@ void InnerLoopVectorizer::vectorizeInterleaveGroup(
+        // %shuffle2 = shuffle %wide.vec, poison, <1, 4, 7, 10> ; 1,4,7,10
+        // %padded = shuffle %wide.vec, poison, <2, 5, 8, 11> ; 2,5,8,11
+        //
+        // %concate1 = shuffle %op1, %op2, <0, 1, ..., 7> ; 0,3,6,9,1,4,7,10
```
should `op1` be `shuffle1` and `op2` be `shuffle2` ?

huhu233 replied:
My apologies... Thanks for the reminder!

nikolaypanchenko reviewed (commented):
Can this be done as a part of VPlan-to-VPlan transformation ? For instance, transform
```
    INTERLEAVE-GROUP with factor 3 at %0, ir<%arrayidx>                                                                                       
      ir<%0> = load from index 0                                                                                                       
      ir<%1> = load from index 1                                                                    
...                               
    INTERLEAVE-GROUP with factor 3 at <badref>, ir<%y>                                       
      store ir<%mul> to index 0                                                                            
      store ir<%mul6> to index 1   
```
into something like
```
    INTERLEAVE-GROUP with factor 3 at %0, ir<%arrayidx>                                                                                       
      ir<%0> = load from index 0                                                                                                       
      ir<%1> = load from index 1                                                                    
      vp<%0> = padded load
...                               
    INTERLEAVE-GROUP with factor 3 at <ir<%0>, ir<%1>, vp<%0>>, ir<%y>                                       
      store ir<%mul> to index 0                                                                            
      store ir<%mul6> to index 1
      store vp<%0> to index 2
```

fhahn commented:
> Can this be done as a part of VPlan-to-VPlan transformation ? For instance, transform

That would be my suggestion as well, the kind of additional maps needed to keep track of additional info is one of the main motivation for VPlan-to-VPlan transformations. This could be a transform replacing a masked interleaved group created originally with an optimized one. I guess the main missing piece is still cost modeling, but I think the main blocker for #67934 has been resolved now so we should be able to make progress on that soon hopefully.

davemgreen commented:
Hello. Just to check - my understand was that we couldn't write back the same value to the missing elements as it would introduce a data race. Is that not correct any more? It might be very unlikely to cause problems in practice.

huhu233 commented:
> Can this be done as a part of VPlan-to-VPlan transformation ? For instance, transform
> 
> ```
>     INTERLEAVE-GROUP with factor 3 at %0, ir<%arrayidx>                                                                                       
>       ir<%0> = load from index 0                                                                                                       
>       ir<%1> = load from index 1                                                                    
> ...                               
>     INTERLEAVE-GROUP with factor 3 at <badref>, ir<%y>                                       
>       store ir<%mul> to index 0                                                                            
>       store ir<%mul6> to index 1   
> ```
> 
> into something like
> 
> ```
>     INTERLEAVE-GROUP with factor 3 at %0, ir<%arrayidx>                                                                                       
>       ir<%0> = load from index 0                                                                                                       
>       ir<%1> = load from index 1                                                                    
>       vp<%0> = padded load
> ...                               
>     INTERLEAVE-GROUP with factor 3 at <ir<%0>, ir<%1>, vp<%0>>, ir<%y>                                       
>       store ir<%mul> to index 0                                                                            
>       store ir<%mul6> to index 1
>       store vp<%0> to index 2
> ```

@nikolaypanchenko @fhahn, thanks for your suggestion, it seems like a more reasonable solution. I'll try to implement it as soon as possible!

huhu233 commented:
> Hello. Just to check - my understand was that we couldn't write back the same value to the missing elements as it would introduce a data race. Is that not correct any more? It might be very unlikely to cause problems in practice.

Hi, @davemgreen, thanks for your comment, it's a good point! But for my understanding, the missing elements is never used in the store interleaved group, will there be a data race scenario?

david-arm commented:
> > Hello. Just to check - my understand was that we couldn't write back the same value to the missing elements as it would introduce a data race. Is that not correct any more? It might be very unlikely to cause problems in practice.
> 
> Hi, @davemgreen, thanks for your comment, it's a good point! But for my understanding, the missing elements is never used in the store interleaved group, will there be a data race scenario?

Looking at this again, I agree with @davemgreen. Unless I've missed something this patch looks unsafe to me - we're deliberately changing the program behaviour. With st3 we are writing to all 3 elements of the struct, but the program only wanted to write the first two elements of each struct.

david-arm commented:
However, it's ok to use ld3 to load just two elements of the struct and discard the 3rd element because that doesn't change memory. In either case I imagine you also need to prove the final 3rd element access of the loop will not seg fault, but I imagine we've already solved that problem.
